src/Catel.Core/Catel.Core.Shared/Reflection/Extensions/MemberInfoExtensions.cs

```diff
@@ -46,6 +48,106 @@ public static string GetSignature(this ConstructorInfo constructorInfo)
         }
 
         /// <summary>
+        /// Constructor distance tuple.
+        /// </summary>
+        private class ConstructorDistance
+        {
+            public ConstructorDistance(int distance, ConstructorInfo constructor)
+            {
+                Distance = distance;
+                Constructor = constructor;
+            }
+
+            public int Distance { get; }
+            public ConstructorInfo Constructor { get; }
+        }
+
+
+        /// <summary>
+        /// Sort constructors
+        /// </summary>
+        /// <param name="constructors"></param>
+        /// <param name="parameters"></param>
+        /// <returns></returns>
+        public static IEnumerable<ConstructorInfo> SortByParametersMatchDistance(this List<ConstructorInfo> constructors, object[] parameters)
+        {
+            List<ConstructorDistance> constructorDistances = new List<ConstructorDistance>();
+            foreach (var constructor in constructors)
+            {
+                int distance;
+                if (constructor.TryGetConstructorDistanceByParametersMatch(parameters, out distance))
+                {
+                    constructorDistances.Add(new ConstructorDistance(distance, constructor));
+                }
+            }
+
+            return constructorDistances.OrderBy(constructor => constructor.Distance).Select(constructor => constructor.Constructor);
+        }
+
+        /// <summary>
+        /// Try to get the constructor distance by parameters match.
+        /// </summary>
+        /// <param name="constructor">The constructor info</param>
+        /// <param name="parameters"></param>
+        /// <param name="distance">The distance</param>
+        /// <returns></returns>
+        public static bool TryGetConstructorDistanceByParametersMatch(this ConstructorInfo constructor, object[] parameters, out int distance)
+        {
+            var constructorParameters = constructor.GetParameters();
+            distance = 0;
+
+            int i = 0;
+            var match = true;
+            while (i < parameters.Length && match)
+            {
+                int value = 0;
+                var parameter = parameters[i];
+                var constructorParameter = constructorParameters[i];
+                var constructorParameterType = constructorParameter.GetType();
+
+                if (parameter == null && !constructorParameterType.IsClassEx() && !constructorParameterType.IsNullableType())
+                {
+                    match = false;
```

Comment:
I think we can also break here? What's the added value of the distance if match == false?